net/ip.c

```diff
@@ -0,0 +1,448 @@
+		rc = na->fmt_pvt.la.nla_autotm ?
+		     snprintf(tmid, ARRAY_SIZE(tmid), "*") :
+		     snprintf(tmid, ARRAY_SIZE(tmid), "%d",
+			      na->fmt_pvt.la.nla_tmid);
+		M0_ASSERT(rc < ARRAY_SIZE(tmid));
+		inet_ntop(AF_INET, &na->ip_n.sn[0], ip_p, ARRAY_SIZE(ip_p));
+		rc = snprintf(buf, M0_NET_IP_STRLEN_MAX,
+			      "%s@%s:12345:%d:%s",
+			      na->fmt_pvt.la.nla_type == M0_NET_IP_PROTO_LO ?
+			      "0" : ip_p,
+			      na->fmt_pvt.la.nla_type ==
+			      M0_NET_IP_PROTO_LO ? "lo" :
+			      ((na->fmt_pvt.la.nla_type ==
+			      M0_NET_IP_PROTO_TCP) ? "tcp": "o2ib"),
+			      na->fmt_pvt.la.nla_portal, tmid);
+	} else if (na->nip_format == M0_NET_IP_INET_IP_FORMAT) {
+		if (na->fmt_pvt.ia.nia_family != M0_NET_IP_AF_UNIX) {
+			inet_ntop(na->fmt_pvt.ia.nia_family ==
+				  M0_NET_IP_AF_INET ? AF_INET : AF_INET6,
+				  &na->ip_n.sn[0], ip_p, ARRAY_SIZE(ip_p));
+			rc = snprintf(buf, M0_NET_IP_STRLEN_MAX, "%s:%s:%s@%d",
+				      na->fmt_pvt.ia.nia_family ==
+				      M0_NET_IP_AF_INET ? "inet" : "inet6",
+				      ip_protocol[na->fmt_pvt.ia.nia_type],
+				      ip_p, na->nip_port);
+		} else
+			M0_LOG(M0_ERROR, "Format is currently not supported");
+	} else if (na->nip_format == M0_NET_IP_INET_HOSTNAME_FORMAT) {
+		inet_ntop(AF_INET, &na->ip_n.sn[0], ip_p, ARRAY_SIZE(ip_p));
+		m0_net_ip_to_hostname(ip_p, hostname);
+		rc = snprintf(buf, M0_NET_IP_STRLEN_MAX, "inet:%s:%s@%d", 
+			      ip_protocol[na->fmt_pvt.ia.nia_type], hostname,
+			      na->nip_port);
+	}
+	if (rc >= M0_NET_IP_STRLEN_MAX)
+		M0_LOG(M0_ERROR, "Name too long. Reqd: %d buf_len: %d", rc,
+		       M0_NET_IP_STRLEN_MAX);
+	M0_LOG(M0_DEBUG, "Address constructed: %s", buf);
+
+	return 0;
+}
+
+M0_INTERNAL int m0_net_hostname_to_ip(char *hostname, char *ip,
+				      enum m0_net_ip_format *fmt)
+{
+	struct hostent  *hname;
+	struct in_addr **addr;
+	uint32_t         ip_n[4];
+	int              i;
+	int              n;
+	char            *cp;
+	char             name[M0_NET_IP_STRLEN_MAX] = {};
+
+	M0_ENTRY("Hostname=%s", (char*)hostname);
+	cp = strchr(hostname, '@');
+	if (cp == NULL)
+		return M0_ERR(-EINVAL);
+
+	n = cp - hostname;
+	name[n] = '\0';
+	strncat(name, hostname, n);
+
+	if (inet_pton(AF_INET, name, &ip_n[0]) == 1 ||
+	    inet_pton(AF_INET6, name, &ip_n[0]) == 1) {
+		/* Copy ip address as it is. */
+		*fmt = M0_NET_IP_INET_IP_FORMAT;
+		strcpy(ip, name);
+	} else {
+		*fmt = M0_NET_IP_INET_HOSTNAME_FORMAT;
+		if ((hname = gethostbyname(name)) == NULL) {
+			M0_LOG(M0_ERROR, "gethostbyname err=%d for %s",
+			       h_errno, (char*)name);
+			/* Return error code for gethostbyname failure */
+			return M0_ERR(h_errno);
+		}
+		addr = (struct in_addr **)hname->h_addr_list;
+		for(i = 0; addr[i] != NULL; i++) {
+			/* Return the first one. */
+			strcpy(ip, inet_ntoa(*addr[i]));
+			M0_LOG(M0_DEBUG, "fqdn=%s ip=%s", (char*)name, ip);
+			return M0_RC(0);
+		}
+
+		/* If no valid addr structure found, then return error */
+		return M0_ERR(-errno);
```

Comment:
This statement means that we called a function that could set `errno`, the function returned -1, and now we want to get the actual error from `errno`. What function did we call? does `gethostbyname` set `errno` in case of success?

Reply:
We have handled possible error conditions before this return statement. If some unknown error comes then we use `errno` to represent error happened during function call.

Reply:
Sorry, I still cannot understand why we are reading `errno` here. Usually we read out `errno` right after a call to a function that sets `errno`. What function sets `errno` in this place?

> If some unknown error comes then we use errno to represent error happened during function call.

Which function call? strchr()? strncat()? inet_pton()? gethostbyname()? we have a lot calls here. Could you please point out what call we aimed at?

Reply:
> Which function call? strchr()? strncat()? inet_pton()? gethostbyname()? 

```
if (inet_pton(AF_INET, name, &ip_n[0]) == 1 ||
	    inet_pton(AF_INET6, name, &ip_n[0]) == 1) {
...
}
```


The `errno` is set by calling above `inet_pton()` in this case.
In case of `gethostbyname()` error will be set in `h_errno` if it fails.